### PR TITLE
[SYCL][Sanitizer] Add fsanitize-target flag for SYCL

### PR DESCRIPTION
**Pull request by jinge90**
"-fsanitize-target=host|device|both" aims to provide mechanism to user to select target to enable address sanitizer in sycl code.
 User can select to only enable Asan pass for device code via "-fsanitize-target=device" or enable for host code only via
 "-fsanitize-target=host". The default value is "both" which means Asan pass will be enabled in both device and host code.
Although this flag is used by sycl for now, it can also be used by openmp offloading if we want to enable sanitizer for it.

### CODE REVIEW

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -733,7 +733,22 @@ static void addSanitizers(const Triple &TargetTriple,
                                          DestructorKind));
       }
     };
-    ASanPass(SanitizerKind::Address, false);
+    // AddressSanitizer can be enabled in offloading scenario to detect bugs in
+    // both host and device code bugs. Sometimes, users may care device or host,
+    // they can use '-fsanitizer-targets' flag to control, the default behavior
+    // is we will enable AddressSanitizer in both host and device compilation.
```
Comment (premanandrao) [inline clang/lib/CodeGen/BackendUtil.cpp:739]:
```suggestion
    // AddressSanitizer can be enabled in offloading scenario to detect bugs in
    // both host and device code.  Users may use '-fsanitizer-targets' flag to enable
    // this only on the host or only on the device.  The default behavior
    // is to enable AddressSanitizer in both host and device compilation.
```

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -733,7 +733,22 @@ static void addSanitizers(const Triple &TargetTriple,
...
+    bool IgnoreAsanPass = false;
+    if (LangOpts.isSYCL()) {
+      auto AsanTarget = CodeGenOpts.getSanitizeTargetsToEnable();
```
Comment (premanandrao) [inline clang/lib/CodeGen/BackendUtil.cpp:742]:
```suggestion
      llvm::AsanTargetsToEnable AsanTarget = CodeGenOpts.getSanitizeTargetsToEnable();
```

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 3]
```diff
@@ -733,7 +733,22 @@ static void addSanitizers(const Triple &TargetTriple,
...
+      if ((AsanTarget == llvm::AsanTargetsToEnable::Device) &&
+          LangOpts.SYCLIsHost)
+        IgnoreAsanPass = true;
+      if ((AsanTarget == llvm::AsanTargetsToEnable::Host) &&
+          LangOpts.SYCLIsDevice)
+        IgnoreAsanPass = true;
+    }
+    if (!IgnoreAsanPass)
+      ASanPass(SanitizerKind::Address, false);
```
Comment (premanandrao) [inline clang/lib/CodeGen/BackendUtil.cpp:751]: Shouldn't this be true?

File: clang/test/Driver/sycl-sanitize-target.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// UNSUPPORTED: system-windows
+
+// RUN: %clangxx -fsycl -fsanitize=address -fsanitize-target=both %s -### 2>&1 \
```
Comment (premanandrao) [inline clang/test/Driver/sycl-sanitize-target.cpp:3]:
```suggestion
// RUN: %clangxx -fsycl -fsanitize=address %s -### 2>&1 \
```
Default is when it is not specified, right?

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 3 above)
Comment (jinge90) [reply clang/lib/CodeGen/BackendUtil.cpp:752]: Hi, @premanandrao 
Could you explain a little more here? What do you think that should be true?
Thanks very much.

Comment (jinge90) [reply clang/lib/CodeGen/BackendUtil.cpp:752]: Hi, @premanandrao 
Kind ping~~
Thanks very much.

Comment (premanandrao) [reply clang/lib/CodeGen/BackendUtil.cpp:752]: This was my misunderstanding about what the second argument meant here.   Sorry.

File: clang/test/Driver/sycl-sanitize-target.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// UNSUPPORTED: system-windows
```
Comment (premanandrao) [inline clang/test/Driver/sycl-sanitize-target.cpp:1]: A brief comment here indicating what the test does will be helpful.

**Review by premanandrao — APPROVED** [review]
CFE changes look good to me.

**Comment by jinge90** [comment]
Hi, @intel/dpcpp-clang-driver-reviewers and @intel/dpcpp-tools-reviewers 
Could you help review this PR?
Thanks very much.

File: clang/include/clang/Driver/Options.td [hunk 6]
```diff
@@ -2514,6 +2514,18 @@ def fsanitize_undefined_strip_path_components_EQ : Joined<["-"], "fsanitize-unde
            "when emitting check metadata.">,
   MarshallingInfoInt<CodeGenOpts<"EmitCheckPathComponentsToStrip">, "0", "int">;
 
+  def sanitize_targets_EQ
+      : Joined<["-"], "fsanitize-target=">,
+        MetaVarName<"<target>">,
+        Visibility<[ClangOption, CC1Option]>,
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:2520]: Should this be available for Windows as well?  If so, need to add `CLOption` to the list of `Visibility` values.

Comment (jinge90) [reply clang/include/clang/Driver/Options.td:2520]: Hi, @mdtoguchi 
We only support device sanitizer for Linux target now.
Thanks very much.

**Pull request closed without merging**